Allow overriding remote settings per environment.

Jelle-S opened the pull request:
Allow overriding remote settings per environment in the properties.yml
file. Itmakes sense for some settings to differ between environments.
For example the `cleandir_limit` setting. It makes sense to maintain
only one or two releases on staging to minimize storage, but to maintain
three or even more releases on production to be able to rollback when
things go wrong.

Jelle-S committed:
Allow overriding remote settings per environment.

Allow overriding remote settings per environment in the properties.yml
file. Itmakes sense for some settings to differ between environments.
For example the `cleandir_limit` setting. It makes sense to maintain
only one or two releases on staging to minimize storage, but to maintain
three or even more releases on production to be able to rollback when
things go wrong.

Jelle-S committed:
Added documentation for environment overrides

dubgeiser commented on README.md:
```diff
@@ -153,6 +153,33 @@ remote:
+  # Here you can specify the overrides per evironment. This example uses the
+  # default regex matcher.
+  environment_overrides:
+    # Here you can specify your overrides. The keys are the values to match
```
Seems like a lot of use of "specify" in one form or the other, suggestiong:
```
List your overrides here.
The keys will be matched against the value of `environment_env_var` via `environment_matcher`.
The default properties use regexMatch and thus the keys here are PCRE regexes.
The values per key can be anything set in `remote` and will override that setting.
Only the first match is used (top to bottom).
```

dubgeiser commented on src/Traits/AbstractCommandTrait.php:
```diff
@@ -86,7 +88,62 @@ protected function getRemoteSettings($host, $user, $keyFile, $app, $timestamp =
+     */
+    protected function processEnvironmentOverrides($settings)
+    {
+        $defaults = [
```
$defaults doesn't need to be reinitialized every single call, so it might be made static, but doesn't have to be IMO.
But, it might be a good idea to declare it outside of the method, this would made possible tests able to override/test it a lot more easily.

Still, not a deal breaker as far as I am concerned ;)

dubgeiser commented on src/Traits/AbstractCommandTrait.php:
```diff
@@ -86,7 +88,62 @@ protected function getRemoteSettings($host, $user, $keyFile, $app, $timestamp =
+            return $settings;
+        }
+
+        // Get the first server in the list.
```
Consider ExtractMethod refactoring into `$server = $this->getFirstServer($settings);` which would replace the redundant comment with self-documenting code.

dubgeiser reviewed (approved):
Think some comments are very valid, but no real deal breakers for me.

Jelle-S committed:
Address review remarks.

- Rephrasing of documentation in README.md
- Extracted default environment override settings to a public static
  property
- Extracted `getFirstServer` method

Jelle-S committed:
Small code style fix.
